File: main workshop/services/typescript/driver-matching-service/src/services/DriverMatchingService.ts
import {
  DynamoDBClient,
  ScanCommand,
  UpdateItemCommand,
} from '@aws-sdk/client-dynamodb';
import {
  EventBridgeClient,
  PutEventsCommand,
} from '@aws-sdk/client-eventbridge';
import { captureAWSv3Client } from 'aws-xray-sdk-core';
import type {
  Driver,
  DriverAssignedEvent,
  PriceCalculatedEvent,
} from '../models';

export interface DriverMatchingResult {
  success: boolean;
  rideId?: string;
  availableDriversCount?: number;
  assignedDriverId?: string;
  errorMessage?: string;
}

export class DriverMatchingService {
  private readonly dynamoDb: DynamoDBClient;
  private readonly eventBridge: EventBridgeClient;
  private readonly driversTableName: string;
  private readonly ridesTableName: string;
  private readonly eventBusName: string;

  constructor() {
    this.dynamoDb = captureAWSv3Client(new DynamoDBClient({}));
    this.eventBridge = captureAWSv3Client(new EventBridgeClient({}));
    this.driversTableName = process.env.DRIVERS_TABLE_NAME || 'Drivers';
    this.ridesTableName =
      process.env.RIDES_TABLE_NAME || 'powertools-ride-workshop-Rides';
    this.eventBusName = process.env.EVENT_BUS_NAME || '';
  }

  async processPriceCalculatedEvent(
    priceCalculatedEvent: PriceCalculatedEvent
  ): Promise<DriverMatchingResult> {
    const result: DriverMatchingResult = {
      success: false,
    };

    try {
      if (!priceCalculatedEvent) {
        result.errorMessage = 'Failed to deserialize price calculated event';
        return result;
      }

      return await this.processRideRequest(priceCalculatedEvent);
    } catch (error) {
      result.errorMessage =
        error instanceof Error ? error.message : 'Unknown error';
      throw error;
    }
  }

  private async processRideRequest(
    request: PriceCalculatedEvent
  ): Promise<DriverMatchingResult> {
    const result: DriverMatchingResult = {
      success: false,
      rideId: request.rideId,
    };

    try {
      // Get available drivers
      const availableDrivers = await this.getAvailableDrivers();
      result.availableDriversCount = availableDrivers.length;

      if (availableDrivers.length === 0) {
        await this.updateRideWithDriver(
          request.rideId,
          '',
          'no-driver-available'
        );
        result.success = false;
        result.errorMessage = 'No available drivers';
        return result;
      }

      // Select the first available driver
      const selectedDriver = availableDrivers[0];
      result.assignedDriverId = selectedDriver.driverId;

      // Update driver status to busy
      // Commented out for demo purposes - we don't need to track real status in the DB
      // await this.updateDriverStatus(selectedDriver.driverId, 'busy');

      // Update ride with driver information and status
      await this.updateRideWithDriver(
        request.rideId,
        selectedDriver.driverId,
        'driver-assigned'
      );

      // Send DriverAssignedEvent to payment processor
      const driverAssignedEvent = this.createDriverAssignedEvent(
        request,
        selectedDriver
      );
      await this.sendDriverAssignedEvent(driverAssignedEvent);

      result.success = true;

      return result;
    } catch (error) {
      result.success = false;
      result.errorMessage =
        error instanceof Error ? error.message : 'Unknown error';
      throw error;
    }
  }

  private async getAvailableDrivers(): Promise<Driver[]> {
    // For demo purposes: fetch all drivers without status filter
    const command = new ScanCommand({
      TableName: this.driversTableName,
      // FilterExpression: '#status = :status',
      // ExpressionAttributeNames: {
      //   '#status': 'status',
      // },
      // ExpressionAttributeValues: {
      //   ':status': { S: 'available' },
      // },
    });

    const response = await this.dynamoDb.send(command);

    const drivers: Driver[] = [];
    if (response.Items) {
      for (const item of response.Items) {
        try {
          const driver = this.deserializeDriver(item);
          drivers.push(driver);
        } catch (error) {
          // Skip invalid driver records
        }
      }
    }

    return drivers;
  }

  private async updateDriverStatus(
    driverId: string,
    status: string
  ): Promise<void> {
    const command = new UpdateItemCommand({
      TableName: this.driversTableName,
      Key: {
        driverId: { S: driverId },
      },
      UpdateExpression: 'SET #status = :status, updatedAt = :updatedAt',
      ExpressionAttributeNames: {
        '#status': 'status',
      },
      ExpressionAttributeValues: {
        ':status': { S: status },
        ':updatedAt': { S: new Date().toISOString() },
      },
    });

    await this.dynamoDb.send(command);
  }

  private async updateRideWithDriver(
    rideId: string,
    driverId: string,
    status: string
  ): Promise<void> {
    const command = new UpdateItemCommand({
      TableName: this.ridesTableName,
      Key: {
        rideId: { S: rideId },
      },
      UpdateExpression:
        'SET driverId = :driverId, #status = :status, updatedAt = :updatedAt',
      ExpressionAttributeNames: {
        '#status': 'status',
      },
      ExpressionAttributeValues: {
        ':driverId': { S: driverId },
        ':status': { S: status },
        ':updatedAt': { S: new Date().toISOString() },
      },
    });

    await this.dynamoDb.send(command);
  }

  private async sendDriverAssignedEvent(
    driverAssignedEvent: DriverAssignedEvent
  ): Promise<void> {
    if (!this.eventBusName) {
      return;
    }

    const command = new PutEventsCommand({
      Entries: [
        {
          Source: 'driver-matching-service',
          DetailType: 'DriverAssigned',
          Detail: JSON.stringify(driverAssignedEvent),
          EventBusName: this.eventBusName,
        },
      ],
    });

    const result = await this.eventBridge.send(command);

    const failedEntries = result.Entries?.filter((e: any) => e.ErrorCode);
    if (failedEntries && failedEntries.length > 0) {
      throw new Error(`Failed to send event: ${failedEntries[0].ErrorCode}`);
    }
  }

  private deserializeDriver(item: Record<string, any>): Driver {
    const currentLocation = this.parseLocation(item);

    return {
      driverId: item.driverId?.S || '',
      driverName: this.getDriverName(item),
      currentLocation: currentLocation,
      status: item.status?.S || 'available',
      rating: item.rating?.N ? Number.parseFloat(item.rating.N) : 5.0,
      createdAt: item.createdAt?.S || new Date().toISOString(),
      updatedAt: this.getUpdatedAt(item),
    };
  }

  private parseLocation(item: Record<string, any>): {
    address: string;
    latitude: number;
    longitude: number;
  } {
    // Try currentLocation field first
    if (item.currentLocation?.S) {
      try {
        return JSON.parse(item.currentLocation.S);
      } catch (error) {
        // Fall through to default
      }
    }

    // Try location field
    if (item.location?.S) {
      try {
        return JSON.parse(item.location.S);
      } catch (error) {
        // Fall through to default
      }
    }

    // Return default location
    return {
      address: '',
      latitude: 0,
      longitude: 0,
    };
  }

  private getDriverName(item: Record<string, any>): string {
    if (item.driverName?.S) {
      return item.driverName.S;
    }
    if (item.name?.S) {
      return item.name.S;
    }
    return 'Unknown Driver';
  }

  private getUpdatedAt(item: Record<string, any>): string {
    if (item.updatedAt?.S) {
      return item.updatedAt.S;
    }
    if (item.lastUpdated?.S) {
      return item.lastUpdated.S;
    }
    return new Date().toISOString();
  }

  private createDriverAssignedEvent(
    request: PriceCalculatedEvent,
    driver: Driver
  ): DriverAssignedEvent {
    return {
      eventType: 'DriverAssigned',
      rideId: request.rideId,
      riderId: request.riderId,
      riderName: request.riderName,
      driverId: driver.driverId,
      driverName: driver.driverName,
      estimatedPrice: request.estimatedPrice,
      basePrice: request.basePrice,
      surgeMultiplier: request.surgeMultiplier,
      pickupLocation: request.pickupLocation,
      dropoffLocation: request.dropoffLocation,
      estimatedArrivalMinutes: 0,
      distanceKm: 0,
      paymentMethod: request.paymentMethod,
      timestamp: new Date().toISOString(),
      correlationId: request.correlationId,
    };
  }
}

File: main-workshop/services/dotnet/driver-matching-service/Services/DriverMatchingService.cs
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using Amazon.EventBridge;
using Amazon.EventBridge.Model;
using System.Text.Json;

[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.SystemTextJson.DefaultLambdaJsonSerializer))]

public class DriverMatchingService
{
    private readonly IAmazonDynamoDB _dynamoDb;
    private readonly IAmazonEventBridge _eventBridge;
    private readonly string _driversTableName;
    private readonly string _ridesTableName;
    private readonly string _eventBusName;
    private readonly JsonSerializerOptions _serializerOptions;

    public DriverMatchingService()
    {
        _dynamoDb = new AmazonDynamoDBClient();
        _eventBridge = new AmazonEventBridgeClient();
        _driversTableName = Environment.GetEnvironmentVariable("DRIVERS_TABLE_NAME") ?? "Drivers";
        _ridesTableName = Environment.GetEnvironmentVariable("RIDES_TABLE_NAME") ?? "powertools-ride-workshop-Rides";
        _eventBusName = Environment.GetEnvironmentVariable("EVENT_BUS_NAME") ?? "";
        
        _serializerOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        };
    }
    
    public async Task<DriverMatchingResult> ProcessRideRequestAsync(RideRequest request)
    {
        if (request == null)
        {
            throw new Exception("Failed to deserialize ride request");
        }
        
        return await ProcessRideRequestInternalAsync(request);
    }

    private async Task<DriverMatchingResult> ProcessRideRequestInternalAsync(RideRequest request)
    {
        var result = new DriverMatchingResult
        {
            RideId = request.RideId,
            EstimatedPrice = request.EstimatedPrice
        };

        try
        {
            // Get available drivers
            var availableDrivers = await GetAvailableDriversAsync();
            result.AvailableDriversCount = availableDrivers.Count;

            if (!availableDrivers.Any())
            {
                await UpdateRideWithDriverAsync(request.RideId, "", "no-driver-available");
                result.Success = false;
                result.ErrorMessage = "No available drivers";
                return result;
            }

            var selectedDriver = availableDrivers.First();
            result.AssignedDriverId = selectedDriver.DriverId;
            
            // Update driver status to busy
            // Commented out for demo purposes - we don't need to track real status in the DB
            // await UpdateDriverStatusAsync(selectedDriver.DriverId, "busy");

            // Update ride with driver information and status
            await UpdateRideWithDriverAsync(request.RideId, selectedDriver.DriverId, "driver-assigned");
            
            // Send DriverAssignedEvent to payment processor
            var driverAssignedEvent = CreateDriverAssignedEvent(request, selectedDriver);
            await SendDriverAssignedEventAsync(driverAssignedEvent);

            result.Success = true;
            return result;
        }
        catch (Exception ex)
        {
            result.Success = false;
            result.ErrorMessage = ex.Message;
            throw; // Re-throw to mark message as failed for retry
        }
    }

    private async Task<List<Driver>> GetAvailableDriversAsync()
    {
        // For demo purposes: fetch all drivers without status filter
        var request = new ScanRequest
        {
            TableName = _driversTableName,
            // FilterExpression = "#status = :status",
            // ExpressionAttributeNames = new Dictionary<string, string>
            // {
            //     { "#status", "status" }
            // },
            // ExpressionAttributeValues = new Dictionary<string, AttributeValue>
            // {
            //     { ":status", new AttributeValue { S = "available" } }
            // }
        };

        var response = await _dynamoDb.ScanAsync(request);
            
        var drivers = new List<Driver>();
        foreach (var item in response.Items)
        {
            try
            {
                var driver = DeserializeDriver(item);
                drivers.Add(driver);
            }
            catch (Exception)
            {
                // ignored
            }
        }
            
        return drivers;
    }

    private async Task<bool> UpdateDriverStatusAsync(string driverId, string status)
    {
        var request = new UpdateItemRequest
        {
            TableName = _driversTableName,
            Key = new Dictionary<string, AttributeValue>
            {
                { "driverId", new AttributeValue { S = driverId } }
            },
            UpdateExpression = "SET #status = :status, updatedAt = :updatedAt",
            ExpressionAttributeNames = new Dictionary<string, string>
            {
                { "#status", "status" }
            },
            ExpressionAttributeValues = new Dictionary<string, AttributeValue>
            {
                { ":status", new AttributeValue { S = status } },
                { ":updatedAt", new AttributeValue { S = DateTime.UtcNow.ToString("O") } }
            }
        };

        await _dynamoDb.UpdateItemAsync(request);
        return true;
    }

    private async Task UpdateRideWithDriverAsync(string rideId, string driverId, string status)
    {
        var updateRequest = new UpdateItemRequest
        {
            TableName = _ridesTableName,
            Key = new Dictionary<string, AttributeValue>
            {
                { "rideId", new AttributeValue { S = rideId } }
            },
            UpdateExpression = "SET driverId = :driverId, #status = :status, updatedAt = :updatedAt",
            ExpressionAttributeNames = new Dictionary<string, string>
            {
                { "#status", "status" }
            },
            ExpressionAttributeValues = new Dictionary<string, AttributeValue>
            {
                { ":driverId", new AttributeValue { S = driverId } },
                { ":status", new AttributeValue { S = status } },
                { ":updatedAt", new AttributeValue { S = DateTime.UtcNow.ToString("O") } }
            }
        };

        await _dynamoDb.UpdateItemAsync(updateRequest);
    }

    private async Task SendDriverAssignedEventAsync(DriverAssignedEvent driverAssignedEvent)
    {
        if (string.IsNullOrEmpty(_eventBusName))
        {
            return;
        }

        var eventDetailJson = JsonSerializer.Serialize(driverAssignedEvent, _serializerOptions);

        var putEventsRequest = new PutEventsRequest
        {
            Entries = new List<PutEventsRequestEntry>
            {
                new()
                {
                    Source = "driver-matching-service",
                    DetailType = "DriverAssigned",
                    Detail = eventDetailJson,
                    EventBusName = _eventBusName
                }
            }
        };

        var result = await _eventBridge.PutEventsAsync(putEventsRequest);
            
        var failedEntries = result.Entries.Where(e => !string.IsNullOrEmpty(e.ErrorCode)).ToList();
        if (failedEntries.Any())
        {
            throw new Exception($"Failed to send event: {failedEntries.First().ErrorCode}");
        }
    }

    // Helper methods
    private Driver DeserializeDriver(Dictionary<string, AttributeValue> item)
    {
        var currentLocation = ParseLocation(item);
        
        return new Driver
        {
            DriverId = item["driverId"].S,
            DriverName = GetDriverName(item),
            CurrentLocation = currentLocation,
            Status = item["status"].S,
            Rating = item.ContainsKey("rating") ? double.Parse(item["rating"].N) : 5.0,
            CreatedAt = item.ContainsKey("createdAt") ? item["createdAt"].S : DateTime.UtcNow.ToString("O"),
            UpdatedAt = GetUpdatedAt(item)
        };
    }

    private Location ParseLocation(Dictionary<string, AttributeValue> item)
    {
        if (item.ContainsKey("currentLocation") && !string.IsNullOrEmpty(item["currentLocation"].S))
        {
            try
            {
                return JsonSerializer.Deserialize<Location>(item["currentLocation"].S) ?? new Location();
            }
            catch (JsonException)
            {
                return new Location();
            }
        }
        
        if (item.ContainsKey("location") && !string.IsNullOrEmpty(item["location"].S))
        {
            try
            {
                return JsonSerializer.Deserialize<Location>(item["location"].S) ?? new Location();
            }
            catch (JsonException)
            {
                return new Location();
            }
        }
        
        return new Location();
    }

    private string GetDriverName(Dictionary<string, AttributeValue> item)
    {
        if (item.ContainsKey("driverName"))
            return item["driverName"].S;
        if (item.ContainsKey("name"))
            return item["name"].S;
        return "Unknown Driver";
    }

    private string GetUpdatedAt(Dictionary<string, AttributeValue> item)
    {
        if (item.ContainsKey("updatedAt"))
            return item["updatedAt"].S;
        if (item.ContainsKey("lastUpdated"))
            return item["lastUpdated"].S;
        return DateTime.UtcNow.ToString("O");
    }

    private DriverAssignedEvent CreateDriverAssignedEvent(RideRequest request, Driver driver)
    {
        return new DriverAssignedEvent
        {
            RideId = request.RideId,
            RiderId = request.RiderId,
            RiderName = request.RiderName,
            DriverId = driver.DriverId,
            DriverName = driver.DriverName,
            EstimatedPrice = request.EstimatedPrice,
            BasePrice = request.BasePrice,
            SurgeMultiplier = request.SurgeMultiplier,
            PickupLocation = request.PickupLocation,
            DropoffLocation = request.DropoffLocation,
            EstimatedArrivalMinutes = 0,
            DistanceKm = 0,
            PaymentMethod = request.PaymentMethod,
            Timestamp = DateTime.UtcNow,
            CorrelationId = request.CorrelationId,
        };
    }
}